tree/ntuple/v7/src/RNTupleMerger.cxx

```diff
@@ -22,10 +22,44 @@
 
 Long64_t ROOT::Experimental::RNTuple::Merge(TCollection *inputs, TFileMergeInfo *mergeInfo)
 {
-   if (inputs == nullptr || mergeInfo == nullptr) {
-      return -1;
+   // Check the inputs
+   if (!inputs || inputs->GetEntries() < 3 || !mergeInfo) return -1;
+
+   // Input parameters
+   std::string ntupleName;
+   std::unique_ptr<Detail::RPageSink> destination;
+   std::vector<std::unique_ptr<Detail::RPageSource>> sources;
+   std::vector<Detail::RPageSource *> sourcePtrs;
+
+   // Parse the inputs
+   Long64_t counter {0};
+   TIter itr(inputs);
+   while(const auto& pitr = itr()) {
+      if ( counter == 0 ) {
+         ntupleName = std::string(pitr->GetName());
+      }
+      else if ( counter == 1 ) {
+         RNTupleWriteOptions writeOpts;
+         writeOpts.SetUseBufferedWrite(false);
+         // THIS WAY CORRUPTS THE OUTPUT FILE!!!
+         destination = Detail::RPageSink::Create(ntupleName, std::string(pitr->GetName()), writeOpts);
+      }
+      else {
+         sources.push_back(Detail::RPageSource::Create(ntupleName, std::string(pitr->GetName())));
```

Comment:
Needless to say, this is not the best way of doing this. On the `TFileMerger` side we seem to have access to the `TDirectory` in which the merged `RNTuple` should be saved and the `TFile` for input files. For now, I only pass the file names and use these in the `RPageSource::Create` and `RPageSink::Create`. The APIs don't seem to support passing reference to existing files, unlike for example `RNTupleWriter`, but only the paths. Then they take care of opening/closing the files but that's not what we want w/ `TFileMerger`.

Having said all this, if one simple takes this toy implementation and pipes the output to a 3rd file (via changing the name of the sink path), it demonstrates that the merging goes through OK via:

```
hadd output.root input[...].root
```

What would be the best way to attach the source/sink to existing files/folders in this case?

Reply:
I think the trick is to directly create an `RPage[Sink|Source]File` object. The file page source can be created from an `RNTuple` anchor object. The file sink can be created with a `TFile` reference using the `Append` factory method.

Eventually I think we want a factory method for the file sink that works with an `std::shared_ptr<TDirectory>`. But for the time being, we may be able to get the corresponding `TFile` reference during merging and accept that we can't handle RNTuples that are not in the files root directory.

Reply:
Thanks a lot Jakob. I now switched to creating the `RPageSource` through the `RNTuple` anchor from the input file. For the output, I switched to using `RPageSinkFile`, where I pass the `TFile` reference. But there is a catch. I see there is already an `RNTuple` anchor in the output file (presumably added through `TFileMerger`). If I use the same `RNTuple` name for the `RPageSinkFile`, then the output writing looks OK but reading fails while unzipping the header. If I use some other name, e.g., name+"_merged" (as currently implemented), all works fine:

```
hadd -ff merged.root input1.root input2.root
```

now gives:

```
$ root -l /tmp/amete/merged.root 
root [0] 
Attaching file /tmp/amete/merged.root as _file0...
(TFile *) 0x3df4140
root [1] .ls
TFile**         /tmp/amete/merged.root
 TFile*         /tmp/amete/merged.root
  KEY: ROOT::Experimental::RNTuple      ntuple_merged;1 object title
  KEY: ROOT::Experimental::RNTuple      ntuple;1        object title
```

`ntuple_merged` looks OK but the `ntuple` seems to be corrupt.

I wonder if the anchor can/should create a sink to which we write to avoid this? Any ideas?

Reply:
> I see there is already an `RNTuple` anchor in the output file (presumably added through `TFileMerger`).

Yes indeed; I was able to get rid of that with the following change:
```diff
diff --git a/io/io/src/TFileMerger.cxx b/io/io/src/TFileMerger.cxx
index 067974b7ed..167aba3de6 100644
--- a/io/io/src/TFileMerger.cxx
+++ b/io/io/src/TFileMerger.cxx
@@ -791,6 +791,9 @@ Bool_t TFileMerger::MergeOne(TDirectory *target, TList *sourcelist, Int_t type,
          ndir->ResetBit(kMustCleanup);
          delete ndir;
       }
+   } else if (std::string(keyclassname) == "ROOT::Experimental::RNTuple") {
+      // The RNTupleMerger has already written the anchor object, do not
+      // overwrite it with bogus data.
    } else if (!canBeFound) { // Don't write the partial result for TTree and TH1
 
       if (!canBeMerged) {
```

Alternatively `TFileMerger` expects that `obj` holds the "merged histogram". In our case, it's still the anchor from the first source file with seek offsets and checksums that don't match the output file. But we can also do something of sorts:
```diff
diff --git a/tree/ntuple/v7/src/RNTupleMerger.cxx b/tree/ntuple/v7/src/RNTupleMerger.cxx
index 0bd6ceff0b..c97c8acd2a 100644
--- a/tree/ntuple/v7/src/RNTupleMerger.cxx
+++ b/tree/ntuple/v7/src/RNTupleMerger.cxx
@@ -29,6 +29,7 @@ Long64_t ROOT::Experimental::RNTuple::Merge(TCollection *inputs, TFileMergeInfo
 
    // Input parameters
    std::string ntupleName;
+   TFile *outFile;
    std::unique_ptr<Internal::RPageSinkFile> destination;
    std::vector<std::unique_ptr<Internal::RPageSource>> sources;
    std::vector<Internal::RPageSource *> sourcePtrs;
@@ -41,7 +42,7 @@ Long64_t ROOT::Experimental::RNTuple::Merge(TCollection *inputs, TFileMergeInfo
          ntupleName = std::string(pitr->GetName());
       }
       else if ( counter == 1 ) {
-         TFile* outFile = pitr->InheritsFrom("TFile") ? dynamic_cast<TFile*>(pitr) : nullptr;
+         outFile = pitr->InheritsFrom("TFile") ? dynamic_cast<TFile*>(pitr) : nullptr;
          if(!outFile) return -1;
          RNTupleWriteOptions writeOpts;
          writeOpts.SetUseBufferedWrite(false);
@@ -65,6 +66,10 @@ Long64_t ROOT::Experimental::RNTuple::Merge(TCollection *inputs, TFileMergeInfo
    Internal::RNTupleMerger merger;
    merger.Merge(sourcePtrs, *destination);
 
+   // Provide the caller with a merged anchor object (even though we've already
+   // written it).
+   *this = *outFile->Get<RNTuple>(ntupleName.c_str());
+
    return 0;
 }
```

Both variants seem to yield files with `RNTuple` anchors that I can read in and process...

Reply:
> Alternatively TFileMerger expects that obj holds the "merged histogram". In our case, it's still the anchor from the first source file 

Why is that?  Doesn't that make the code harder to follow if `obj` has a different semantic?  

```
   } else if (std::string(keyclassname) == "ROOT::Experimental::RNTuple") {
          ...
   } else if (!canBeFound) { // Don't write the partial result for TTree and TH1
```
That sort-of make sense since `canBeFound` is not true for `RNTuple` because, not being a `TObject` and not being shared owned by the `TFile`, it is not in the list of connected objects (`TFileMerger.cxx:545).


Instead of:
```
   *this = *outFile->Get<RNTuple>(ntupleName.c_str());
```
`TTree` uses the similar (but once less copy):
```
   info->fOutputDirectory->ReadTObject(this,this->GetName());
```
which does not yet exist for non TObject but could implemented easily (same implementation as `ReadTObject` but using `TKey::ReadObjWithBuffer`).

Reply:
Thanks a lot @hahnjo, @pcanal. I opted for the first option in the latest commit, primarily due to its clarity/brevity, and indeed with this change things seem to work as expected. Please let me know how you'd like to proceed on this.

Reply:
> > Alternatively TFileMerger expects that obj holds the "merged histogram". In our case, it's still the anchor from the first source file
> 
> Why is that? Doesn't that make the code harder to follow if `obj` has a different semantic?

I wasn't precise here: `obj` / `this` is a *copy* of the anchor from the first source file that will eventually end up in the destination file. And indeed, I was quickly chatting with Jakob about this yesterday and we were leaning towards avoiding special cases and different semantics as much as possible, ie rather implementing the second approach.

> Instead of:
> 
> ```
>    *this = *outFile->Get<RNTuple>(ntupleName.c_str());
> ```
> 
> `TTree` uses the similar (but once less copy):
> 
> ```
>    info->fOutputDirectory->ReadTObject(this,this->GetName());
> ```
> 
> which does not yet exist for non TObject but could implemented easily (same implementation as `ReadTObject` but using `TKey::ReadObjWithBuffer`).

Yes, that's why I wrote "something of sorts"; ideally we would want the `RPageSinkFile` to give access to the `RNTuple` anchor object it just wrote...

Reply:
I now adopted the second approach, i.e., overwriting the `RNTuple` anchor on the `RNTupleMerger` side with no special treatment for `RNTuple` on the `TFileMerger` side.